File: handyman.parent/handyman/src/main/scala/in/handyman/process/onethread/CopyDataJdbcWriter.scala
package in.handyman.process.onethread

import java.util.concurrent.BlockingQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Callable
import com.typesafe.scalalogging.LazyLogging
import net.sf.jsqlparser.statement.insert.Insert
import scala.collection.mutable.HashSet
import in.handyman.util.ResourceAccess
import in.handyman.HandymanException
import scala.util.control.Breaks
import java.sql.SQLException
import java.sql.Statement
import java.util.concurrent.atomic.AtomicInteger

class CopyDataJdbcWriter(configMap: Map[String, String], insert: Insert, poisonPill: Row,
                         copyData: in.handyman.dsl.Copydata,
                         id:       String, rowQueue: BlockingQueue[Row],
                         countDownLatch: CountDownLatch) extends Callable[Void] with LazyLogging {

  val writeBuffer: HashSet[String] = new HashSet[String]
  val target = {
    if (!copyData.getTo.trim.isEmpty())
      copyData.getTo.trim
    else
      throw new HandymanException("target data source cannot be empty for copydata for " + copyData.getName)
  }

  val conn = {
    val c = ResourceAccess.rdbmsConn(target)
    c.setAutoCommit(false)
    c
  }

  val writeSize = {
    if (!copyData.getWriteBatchSize.isEmpty && copyData.getWriteBatchSize.toInt > 0)
      copyData.getWriteBatchSize.toInt
    else {
      configMap.getOrElse(Constants.WRITESIZE, Constants.DEFAULT_WRITE_SIZE).toInt
    }
  }

  val columnList = {
    val colListBuilder = new StringBuilder
    val colListSize = insert.getColumns.size
    insert.getColumns.forEach(item => {
      val currentIndex = insert.getColumns.indexOf(item)
      if (currentIndex + 1 == colListSize)
        colListBuilder.append(item.getColumnName)
      else
        colListBuilder.append(item.getColumnName).append(",")
    })
    colListBuilder.toString
  }

  def call(): Void = {

    Breaks.breakable {
      while (true) {
        val row = rowQueue.take();
        if (poisonPill.equals(row)) {
          if (!writeBuffer.isEmpty) {
            logger.info(s"CopydataWriter(After poison pill) flushing to database rows:$writeBuffer.size")
            writeToDb
          }
          countDownLatch.countDown
          Breaks.break

        } else {
          val dataFrame = generateDataFrame(row)
          writeBuffer.add(dataFrame)
          if (writeBuffer.size % writeSize == 0) {
            logger.info(s"CopydataWriter(Before poison pill) flushing to database rows:$writeBuffer.size")
            writeToDb
          }
        }
      }
    }

    ???
  }

  def generateDataFrame(row: Row): String = {
    logger.debug(s"Copydata Writer generating dataframe for row:$row")
    val columnSet = row.columnSet
    val dataFrameBuilder = new StringBuilder

    val targetTable = insert.getTable

    dataFrameBuilder.append("INSERT INTO ").append(targetTable).append(" (").
      append(columnList).append(") VALUES").append(Constants.INSERT_STMT_VALUE_START)

    columnSet.foreach(column => {
      val columnType = column.columnTypeName
      columnType.toLowerCase match {
        case Constants.STRING_DATATYPE => dataFrameBuilder.append(Constants.STRING_ENCLOSER).
          append(column.value).append(Constants.STRING_ENCLOSER)
        case "datetime" => dataFrameBuilder.append(Constants.STRING_ENCLOSER).
          append(column.value).append(Constants.STRING_ENCLOSER)
        case "timestamp" => dataFrameBuilder.append(Constants.STRING_ENCLOSER).
          append(column.value).append(Constants.STRING_ENCLOSER)
        case _ => dataFrameBuilder.append(column.value)
      }
      if (!column.isLastColumn)
        dataFrameBuilder.append(Constants.FIELD_SEPARATOR)
    })

    dataFrameBuilder.append(Constants.INSERT_STMT_VALUE_END)
    dataFrameBuilder.toString()
  }

  def writeToDb() = {
    logger.info(s"Writing to database using conn:$target")
    val stmt: Statement = conn.createStatement
    try {
      writeBuffer.foreach(sql => {
        stmt.addBatch(sql)
      })
      stmt.executeBatch
      conn.commit
      writeBuffer.clear
    } catch {
      case ex: SQLException => {
        logger.error(s"CopydataWriter:$id error closing source connection for database:$target", ex)
      }
      case ex: Throwable => {
        logger.error(s"CopydataWriter:$id error closing source connection for database:$target", ex)
      }
    } finally {
      try {
        if (stmt != null)
          stmt.close
      } catch {
        case ex: Throwable => {
          ex.printStackTrace
          logger.error(s"Copydata:$id error closing source connection for database:$target", ex)
        }
      }
    }

  }

  def cleanup = {
    /*try {
      if (conn != null) {
        conn.close()
      }
    } catch {
      case ex: Exception => {
        ex.printStackTrace()
        //throw ex
      }
    }*/
  }
}

File: handyman.raven/src/main/java/in/handyman/raven/lib/CopyDataJdbcWriter.java
package in.handyman.raven.lib;

import in.handyman.raven.connection.ResourceAccess;
import in.handyman.raven.exception.HandymanException;
import in.handyman.raven.lib.model.CopyData;
import in.handyman.raven.lambda.doa.Action;
import in.handyman.raven.lambda.process.LambdaEngine;
import in.handyman.raven.util.Table;
import in.handyman.raven.util.UniqueID;
import net.sf.jsqlparser.schema.Column;
import net.sf.jsqlparser.statement.insert.Insert;
import org.slf4j.Logger;

import java.sql.Connection;
import java.sql.Statement;
import java.util.HashSet;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.stream.Collectors;

public class CopyDataJdbcWriter implements Callable<Void> {

    private final Insert insert;
    private final Table.Row poisonPill;
    private final Action action;
    private final LinkedBlockingDeque<Table.Row> rowQueue;
    private final CountDownLatch countDownLatch;

    private final Set<String> writeBuffer = new HashSet<>();
    private final String target;
    private final Integer writeSize;
    private final String columnList;
    private final Logger log;

    public CopyDataJdbcWriter(final Map<String, String> configMap, final Insert insert,
                              final Table.Row poisonPill, final CopyData copyData, final Action action,
                              final LinkedBlockingDeque<Table.Row> rowQueue, final CountDownLatch countDownLatch) {
        this.insert = insert;
        this.poisonPill = poisonPill;
        this.action = action;
        this.rowQueue = rowQueue;
        this.countDownLatch = countDownLatch;

        this.target = Optional.ofNullable(copyData.getTo()).map(String::trim)
                .filter(s -> !s.isEmpty() && !s.isBlank())
                .orElseThrow(() -> new HandymanException("target data source cannot be empty for copyData for " + copyData.getName()));
        this.writeSize = Optional.ofNullable(copyData.getWriteBatchSize())
                .map(String::trim)
                .map(Integer::valueOf)
                .filter(integer -> integer > 0)
                .orElseGet(() -> Integer.valueOf(configMap.getOrDefault(Constants.WRITE_SIZE, Constants.DEFAULT_WRITE_SIZE).trim()));
        this.columnList = insert.getColumns().stream().map(Column::getColumnName).collect(Collectors.joining(","));

        this.log = LambdaEngine.getLogger(action);
    }


    @Override
    public Void call() throws Exception {

        while (true) {
            final Table.Row row = rowQueue.take();
            if (poisonPill.equals(row)) {
                if (!writeBuffer.isEmpty()) {
                    log.info("CopyDataWriter(After poison pill) flushing to database rows:{}", writeBuffer.size());
                    writeToDb();
                }
                countDownLatch.countDown();
                break;
            } else {
                final String dataFrame = generateDataFrame(row);
                if (Objects.nonNull(dataFrame)) {
                    writeBuffer.add(dataFrame);
                    if (writeBuffer.size() % writeSize == 0) {
                        log.info("CopyDataWriter(Before poison pill) flushing to database rows:{}", writeBuffer.size());
                        writeToDb();
                    }
                }
            }
        }

        return null;
    }

    private void writeToDb() {

        var hikariDataSource = ResourceAccess.rdbmsConn(target);
        try (final Connection sourceConnection = hikariDataSource.getConnection()) {
            sourceConnection.setAutoCommit(false);
            log.info("Writing to database using conn: {}", target);
            final Long statementId = UniqueID.getId();
            //TODO audit
            try (final Statement stmt = sourceConnection.createStatement()) {
                for (var s : writeBuffer) {
                    stmt.addBatch(s);
                }
                stmt.executeBatch();
                sourceConnection.commit();
                writeBuffer.clear();
            }
        } catch (Throwable ex) {
            log.error("CopyDataWriter: {} error closing source connection for database: {} ", action.getActionId(), target, ex);
            throw new HandymanException("writeToDb failed", ex);
        }
    }

    private String generateDataFrame(final Table.Row row) {

        log.info("CopyDataWriter: Writer generating dataframe for row: {}", row);
        var columnSet = row.getColumnSet();
        var dataFrameBuilder = new StringBuilder();

        var targetTable = insert.getTable();

        dataFrameBuilder.append("INSERT INTO ").append(targetTable).append(" (").
                append(columnList).append(") VALUES").append(Constants.INSERT_STMT_VALUE_START);

        if (Objects.nonNull(columnSet)) {
            for (final Table.ColumnInARow column : columnSet) {
                final String columnTypeName = column.getColumnTypeName();

                if (Objects.nonNull(columnTypeName) && (Objects.equals(columnTypeName.toLowerCase(), Constants.STRING_DATATYPE)
                        || Objects.equals(columnTypeName.toLowerCase(), "java.lang.string")
                        || Objects.equals(columnTypeName.toLowerCase(), "datetime")
                        || Objects.equals(columnTypeName.toLowerCase(), "timestamp"))) {
                    dataFrameBuilder.append(Constants.STRING_ENCLOSER).
                            append(column.getValue()).append(Constants.STRING_ENCLOSER);
                } else {
                    dataFrameBuilder.append(column.getValue());
                }
                if (!column.getIsLastColumn()) {
                    dataFrameBuilder.append(Constants.FIELD_SEPARATOR);
                }
            }
            dataFrameBuilder.append(Constants.INSERT_STMT_VALUE_END);
            return dataFrameBuilder.toString();

        }
        return null;
    }

}
